core/state/state_object.go

```diff
@@ -233,7 +240,9 @@ func (c *StateObject) AddBalance(amount *big.Int) {
 }
 
 func (c *StateObject) SubBalance(amount *big.Int) {
-	if amount.Cmp(common.Big0) == 0 {
+	// We must check emptiness for the objecst such that the account
+	// clearing (0,0,0 objects) can take effect.
+	if amount.Cmp(common.Big0) == 0 && c.empty() {
```

Comment:
Shouldn't this be `!c.empty()`. But I'm not sure really about the case you're coding for here... 

1. During init-code, if the incoming balance is sent off. But then the nonce is already incremented, right? So that can't happen.
2. During suicide, subbalance may be called. And then we could safely return here (?)
3. In case it's a normal account which does the sending, and becomes empty, wouldn't the nonce already be incremented at this point? If so, the `c.empty()` cannot happen

So is it case 2 ? In that case I guess it's correct, but you maybe should document the reason for this check.